### PR TITLE
Metadata intake

### PR DESCRIPTION
**Pull request by lucasortizny**
Here is my pull request for the metadata intake. Some notes to recap: 

- The implementation in server.py is in handle_setup_v01. As requested, handle_setup has been left intact.
- The master table has two new columns: participant_id (unique, nullable, max size is 50 characters in length, auto-generated UUID4 if not specified) and extra_metadata (4096 characters in length).
- The extra metadata information will be serialized/deserialized -> JSON and placed into the Database in that way for future reference. 
- The Database schematic I originally created will need to be updated with the new information. 

Please let me know if there are any questions! This should help cover #32

### CODE REVIEW

File: aepsych/config.py
Comment (mshvartsman) [inline aepsych/config.py:107]: Is that right? The way ConfigParser works, things that are in `common` are mirrored to all other sections. What this method is doing (I think) is omitting these mirrored copies (basically if something is in `common`, and also shows up in another section, just ignore the one in the other section since it's a mirror).

File: aepsych/database/db.py
Comment (mshvartsman) [inline aepsych/database/db.py:152]: can you underscore `extra_metadata`?

File: aepsych/server/server.py
Comment (mshvartsman) [inline aepsych/server/server.py:353]: shouldn't this be `experiment_name` and `experiment_description` instead of these hardcoded defaults?

File: aepsych/server/server.py
Comment (mshvartsman) [inline aepsych/server/server.py:338]: Why not modify `self.configure` to take in a `Config` object instead of the raw message? Then you don't need to create it twice (the first thing `self.configure` does is create a `Config` anyway).

**Comment by mshvartsman** [comment]
Great start -- left some nits in the code. Can you put together a minimal example that constructs some messages with metadata, processes them through `versioned_handler`, and then gets that metadata back out? That would be good as a test, and also as an example for users. 

Also, once #89 lands you should run `ufmt` on your changes.

File: aepsych/server/server.py
Comment (lucasortizny) [reply aepsych/server/server.py:338]: Hey @mshvartsman ! I absolutely agree in that it would be a great idea to just pass the config object. One thing I am concerned about is that self.configure is also called by the legacy `self.configure` function with a string. I would have to write another version of this function to adjust to the new `handle_setup_v01` but again it is important to think about whether you'd like to see another secondary version of an existing function. Please let me know if you'd like me to proceed. (TLDR: I don't want to modify the original `handle_setup` in any way unless absolutely necessary).

File: aepsych/config.py
Comment (lucasortizny) [reply aepsych/config.py:107]: Oh I absolutely did not know that. I knew default sections in ConfigParsers were special but now it makes sense why this line is here!

File: aepsych/server/server.py
Comment (mshvartsman) [reply aepsych/server/server.py:338]: I think modifying the original `handle_setup` is maybe okay here. To the extent that there's a contract here it's that legacy `handle_setup`'s behavior doesn't change, which will still be true if you change where the string-to-config conversion happens. I don't think anything else is calling `self.configure`. @crasanders what do you think?

File: aepsych/database/tables.py
Comment (crasanders) [inline aepsych/database/tables.py:64]: Do we want participant_id to be nullable?

File: aepsych/server/server.py
Comment (crasanders) [reply aepsych/server/server.py:338]: Yeah, I think it's fine to change handle_setup to pass a `Config` into `self.configure`.

File: aepsych/server/server.py
Comment (crasanders) [reply aepsych/server/server.py:414]: I don't have an opinion one way or the other about capitalization, but do make sure that it's consistent. It might be useful to define `DEFAULT_NAME` and `DEFAULT_DESC` constants.

File: aepsych/database/tables.py
Comment (lucasortizny) [reply aepsych/database/tables.py:62]: I originally wanted it that way but then I eventually wrote it so a UUID would be generated anyways so this line no longer makes sense so I will remove the nullability!

File: aepsych/server/server.py
Comment (lucasortizny) [reply aepsych/server/server.py:338]: Okie dokie, will do!

File: aepsych/config.py
Comment (mshvartsman) [reply aepsych/config.py:107]: Ok -- can you update the comment to be correct or remove?

Comment (lucasortizny) [reply aepsych/config.py:107]: Yeah I'll update it, haven't pushed any commits yet

File: aepsych/server/server.py
Comment (mshvartsman) [reply aepsych/server/server.py:353]: Yeah I'm not sure what I was thinking. If there's no metadata we can keep it to some default strung.

File: aepsych/server/server.py
Comment (mshvartsman) [reply aepsych/server/server.py:414]: Good idea

**Comment by lucasortizny** [comment]
Attached is the current Database schematic updated to this PR's changes. I am going to continue this PR by writing unit tests for the new functionality in the DB. 
Database Mapping July 14th

**Comment by lucasortizny** [comment]
I don't know what other kind of unit tests I can write (unit testing is not my strong suit haha). Would it be worth to write unit tests for the new implementation of the configure command as well? I'm not sure if it will be necessary considering it still maintains backwards compatibility with legacy versions of aepsych

**Comment by lucasortizny** [comment]
Want to follow up, did some more research on unit testing and its purpose and I have an idea of what kind of tests I can come up with!

File: aepsych/config.py
Comment (crasanders) [inline aepsych/config.py:109]: It looks like this was copied from `__str__`, which is fine, but maybe this implies that we should have a helper function that both methods use so that we're not duplicating code.

File: aepsych/database/tables.py
Comment (crasanders) [inline aepsych/database/tables.py:123]: This code is redundant with `_has_participant_id`. Do you even need to run these checks since you already call `_has_participant_id` on line 93?

File: aepsych/database/tables.py
Comment (crasanders) [inline aepsych/database/tables.py:141]: Similar comment as above.

File: aepsych/server/server.py
Comment (crasanders) [inline aepsych/server/server.py:336]: Will this break if there's no `"config_str"` or `"config_dict"` in the message (we do checks for this in line 373)?

File: aepsych/server/server.py
Comment (crasanders) [inline aepsych/server/server.py:453]: The old way should still work, right? So is this change necessary?

File: aepsych/server/server.py
Comment (crasanders) [inline aepsych/server/server.py:720]: I don't like "configasobject" as a parameter name. Could we just call it "config"?

File: tests/test_db.py
Comment (crasanders) [inline tests/test_db.py:246]: It looks like what this test does is creates a Config and configures a server using the same config string, and then verifies that the Config and the server database contain the same data. I would like to see a more explicit check that the database has correctly recorded `experiment_name = Lucas`, `metadata1 = one`, etc. There should also be checks to see what happens if metadata is malformed or missing (e.g., what happens if we don't supply `experiment_name`)? We should also probably add a test or two to the config tests to makes sure stuff is jsonifying correctly.

**Review by crasanders — CHANGES_REQUESTED** [review]

File: aepsych/database/tables.py
Comment (lucasortizny) [reply aepsych/database/tables.py:123]: Considering participant_id column will be added at the same time as the metadata column, you are right in that it would be able to infer that if there is no participant_id column, there is no metadata column. But what about (for edge cases should this ever be a situation) if there was a participant_id column but no metadata column? Shouldn't we try to take into account any potential edge cases like those?

File: aepsych/server/server.py
Comment (lucasortizny) [reply aepsych/server/server.py:453]: Good point! I'll revert it back since it's an unnecessary step. Good catch 👍

File: aepsych/database/tables.py
Comment (lucasortizny) [reply aepsych/database/tables.py:123]: Actually I see what you mean. I should just have one method that checks for a column name based on an input. But I still run the checks individually of each column to handle any extra edge cases

File: aepsych/server/server.py
Comment (lucasortizny) [reply aepsych/server/server.py:336]: @crasanders may I get some clarification on this front?

File: aepsych/server/server.py
Comment (lucasortizny) [reply aepsych/server/server.py]: Agreed

File: tests/test_db.py
Comment (lucasortizny) [reply tests/test_db.py:251]: [NOTE TO SELF FROM MAINTAINER MEETING]
originally check -> config object created but new code flips the result and can cause error. maybe include the code in there to check it?

**Comment by lucasortizny** [comment]
The reason why config was changed to usedconfig is because the metadata-intake functionality introduces a "config" parameter in the configure function in line 677 that I don't want to run into variable naming issues.

**Pull request closed without merging**
